testng-core/src/main/java/org/testng/internal/TestListenerHelper.java

```diff
@@ -196,9 +176,37 @@ public static ITestNGListenerFactory createListenerFactory(
   }
 
   public static class ListenerHolder {
-    private List<Class<? extends ITestNGListener>> listenerClasses;
+    private final List<Class<? extends ITestNGListener>> listenerClasses = new ArrayList<>();
     private Class<? extends ITestNGListenerFactory> listenerFactoryClass;
 
+    @SuppressWarnings("unchecked")
+    public void addListener(Class<? extends ITestNGListener> c) {
+      // @Listener annotation is now inheritable. So let's add it ONLY
+      // if it wasn't added already
+      if (!listenerClasses.contains(c)) {
+        listenerClasses.add(c);
+      }
+      boolean isListenerFactory = ITestNGListenerFactory.class.isAssignableFrom(c);
```

Comment:
The logic if not clear. Please introduce a private setListenerFactory

Reply:
Refactored and pushed. Please check now.